virtdbg/vmx/vmm.c

```diff
@@ -0,0 +1,290 @@
+
+    vmwrite(HOST_GDTR_BASE, (size_t)gdtr.address);
+    vmwrite(HOST_IDTR_BASE, (size_t)idtr.address);
+    //save efer
+    vmwrite(HOST_EFER_FULL, __rdmsr(0xc0000080));
+
+    //Set up guest state on vm entry.
+    //same values as gdt.c
+    vmwrite(GUEST_ES_SELECTOR, state->es);
+    vmwrite(GUEST_CS_SELECTOR, state->cs);
+    vmwrite(GUEST_DS_SELECTOR, state->ds);
+    vmwrite(GUEST_FS_SELECTOR, state->fs);
+    vmwrite(GUEST_GS_SELECTOR, state->gs);
+    vmwrite(GUEST_SS_SELECTOR, state->ss);
+    vmwrite(GUEST_TR_SELECTOR, 0x0);
+    vmwrite(GUEST_LDTR_SELECTOR, 0x0);
+    vmwrite(GUEST_CS_BASE, 0x0);
+    vmwrite(GUEST_DS_BASE, 0x0);
+    vmwrite(GUEST_ES_BASE, 0x0);
+    vmwrite(GUEST_FS_BASE, 0x0);
+    vmwrite(GUEST_GS_BASE, 0x0);
+    vmwrite(GUEST_SS_BASE, 0x0);
+    vmwrite(GUEST_LDTR_BASE, 0x0);
+    vmwrite(GUEST_IDTR_BASE, (size_t)state->idt.address);
+    vmwrite(GUEST_GDTR_BASE, (size_t)state->gdt.address);
+    vmwrite(GUEST_TR_BASE, 0x0);
+    vmwrite(GUEST_CS_LIMIT, 0xffff);
+    vmwrite(GUEST_DS_LIMIT, 0xffff);
+    vmwrite(GUEST_ES_LIMIT, 0xffff);
+    vmwrite(GUEST_FS_LIMIT, 0xffff);
+    vmwrite(GUEST_GS_LIMIT, 0xffff);
+    vmwrite(GUEST_SS_LIMIT, 0xffff);
+    vmwrite(GUEST_LDTR_LIMIT, 0xffff);
+    vmwrite(GUEST_TR_LIMIT, 0xffff);
+    vmwrite(GUEST_GDTR_LIMIT, state->gdt.size);
+    vmwrite(GUEST_IDTR_LIMIT, state->idt.size);
+
+    vmwrite(GUEST_CS_ACCESS_RIGHT, CODE_ACCESS_RIGHT);
+    vmwrite(GUEST_DS_ACCESS_RIGHT, DATA_ACCESS_RIGHT);
+    vmwrite(GUEST_ES_ACCESS_RIGHT, DATA_ACCESS_RIGHT);
+    vmwrite(GUEST_FS_ACCESS_RIGHT, DATA_ACCESS_RIGHT);
+    vmwrite(GUEST_GS_ACCESS_RIGHT, DATA_ACCESS_RIGHT);
+    vmwrite(GUEST_SS_ACCESS_RIGHT, DATA_ACCESS_RIGHT);
+    vmwrite(GUEST_LDTR_ACCESS_RIGHT, LDTR_ACCESS_RIGHT);
+    vmwrite(GUEST_TR_ACCESS_RIGHT, TR_ACCESS_RIGHT);
+    vmwrite(GUEST_DR7, 0x0);
+    vmwrite(GUEST_RFLAG, state->rflags);
+    //used for nested virtualization
+    vmwrite(VMCS_LINK_POINTER, -1ll);
+    vmwrite(VMCS_FIELD_GUEST_EFER_FULL, state->efer);
+
+    vmwrite(GUEST_INTR_STATUS, 0);
+    vmwrite(GUEST_PML_INDEX, 0);
+    //copy fixed values to guest
+    uint64_t cr0FixedGuest = __rdmsr(IA32_VMX_CR0_FIXED0_MSR);
+    uint32_t cr0FixedLo = (uint32_t)cr0FixedGuest;
+    uint32_t cr0FixedHi = (uint32_t)(cr0FixedGuest >> 32);
+    vmwrite(GUEST_CR0, cr0FixedLo | ((uint64_t)cr0FixedHi) << 32 | state->cr0);
+    TRACE("cr3: %X", state->cr3);
+    vmwrite(GUEST_CR3, state->cr3);
+    uint64_t cr4FixedGuest = __rdmsr(IA32_VMX_CR4_FIXED0_MSR);
+    uint32_t cr4FixedLo = (uint32_t)cr4FixedGuest;
+    uint32_t cr4FixedHi = (uint32_t)(cr4FixedGuest >> 32);
+    vmwrite(GUEST_CR4, cr4FixedLo | ((uint64_t)cr4FixedHi) << 32 | state->cr4);
+    vmwrite(GUEST_INTERRUPTIBILITY_STATE, 0x0);
+    vmwrite(GUEST_ACTIVITY_STATE, 0x0);
+
+    vmwrite(CTLS_EPTP, (uintptr_t)g_root_pa | 6 | (4 - 1) << 3);
+
+    vmwrite(GUEST_RSP, state->rsp);
+    vmwrite(GUEST_RIP, state->rip);
+
+    uint16_t es;
```

Comment:
why are you not copying these from the guest state? if these are not there then we should pass them as well